### PR TITLE
Spec func pass

### PR DESCRIPTION
**Pull request by AVassilev98**

### CODE REVIEW

**Comment by AVassilev98** [comment]
commit 1f56057 will break clam since I modified the function signatures of the global analyses, Will make a separate PR for fixing clam.

File: include/seadsa/SpecGraphInfo.hh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+namespace seadsa {
+class LocalAnalysis;
+class AllocWrapInfo;
+
+class SpecGraphInfo : public llvm::ImmutablePass {
+
+protected:
+  mutable AllocWrapInfo *m_awi;
+  mutable llvm::TargetLibraryInfoWrapperPass *m_tliWrapper;
+  mutable std::unordered_map<std::string, llvm::Function &> m_funcs;
```
Comment (caballa) [inline include/seadsa/SpecGraphInfo.hh:36]: maybe better `const llvm::Function*` rather than `llvm::Function&`?

File: lib/seadsa/DsaBottomUp.cc [hunk 2]
```diff
@@ -93,26 +94,25 @@ void BottomUpAnalysis::cloneAndResolveArguments(const DsaCallSite &CS,
...
     Cell c(n, ret.getRawOffset());
     nc.unify(c);
   }
 
-  // clone and unify actuals and formals
   DsaCallSite::const_actual_iterator AI = CS.actual_begin(),
                                      AE = CS.actual_end();
-  for (DsaCallSite::const_formal_iterator FI = CS.formal_begin(),
-                                          FE = CS.formal_end();
+  for (DsaCallSite::const_formal_iterator FI = CS.formal_func_begin(callee),
```
Comment (caballa) [inline lib/seadsa/DsaBottomUp.cc:107]: Does not have `CS` already the callee, why do we need to pass it?

File: lib/seadsa/DsaGlobal.cc [hunk 3]
```diff
@@ -104,8 +105,7 @@ bool ContextInsensitiveGlobalAnalysis::runOnModule(Module &M) {
     // --- all scc members share the same local graph
     for (CallGraphNode *cgn : scc) {
       Function *fn = cgn->getFunction();
-      if (!fn || fn->isDeclaration() || fn->empty())
-        continue;
+      if (!fn || fn->isDeclaration() || fn->empty()) continue;
```
Comment (caballa) [inline lib/seadsa/DsaGlobal.cc:108]: I think we only pass `specGraphInfo` to two analyses:  BottomUpTopDown and ContextSensitive. We should also pass it to all derived classes from `GlobalAnalysis` and `DsaGlobalPass` that includes the context insensitive variant and the flat one.

File: lib/seadsa/SpecGraphInfo.cc [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  if (SpecFile.empty()) return;
+
+  SMDiagnostic err;
+
+  m_module = parseIRFile(SpecFile, err, m_ctx);
+
+  auto &dl = m_module->getDataLayout();
+
+  LocalAnalysis la(dl, *m_tliWrapper, *m_awi);
```
Comment (caballa) [inline lib/seadsa/SpecGraphInfo.cc:52]: `la` seems unused

File: lib/seadsa/DsaGlobal.cc (hunk 3 above)
Comment (caballa) [reply lib/seadsa/DsaGlobal.cc:108]: Also `CompleteCallGraphAnalysis` would benefit from `specGraphInfo`

File: lib/seadsa/DsaBottomUp.cc (hunk 2 above)
Comment (AVassilev98) [reply lib/seadsa/DsaBottomUp.cc:107]: we pass in the spec function if we have it, otherwise we use the callee that CS has (see line 86)

File: lib/seadsa/SpecGraphInfo.cc (hunk 4 above)
Comment (AVassilev98) [reply lib/seadsa/SpecGraphInfo.cc:52]: must have missed it, will remove

File: lib/seadsa/DsaBottomUp.cc (hunk 2 above)
Comment (AVassilev98) [reply lib/seadsa/DsaBottomUp.cc:107]: we can't use the formals from the CS callee because `calleeG.hasCell(*fml)` would always fail if calleeG was generated using a spec function

File: include/seadsa/SpecGraphInfo.hh (hunk 1 above)
Comment (AVassilev98) [reply include/seadsa/SpecGraphInfo.hh:36]: I can do this, but I'm not sure what the reasoning behind this is?

File: lib/seadsa/SpecGraphInfo.cc (hunk 4 above)
Comment (AVassilev98) [reply lib/seadsa/SpecGraphInfo.cc]: 99cdaf1

File: include/seadsa/SpecGraphInfo.hh (hunk 1 above)
Comment (AVassilev98) [reply include/seadsa/SpecGraphInfo.hh]: can't have it const because of the call [here](https://github.com/seahorn/sea-dsa/blob/02a7dfb09068a27d1261fbc115a98dec0a915cb5/lib/seadsa/DsaBottomUp.cc#L148), but I agree that having it as pointer type is a better choice. Change here: 2245707

File: lib/seadsa/DsaGlobal.cc (hunk 3 above)
Comment (AVassilev98) [reply lib/seadsa/DsaGlobal.cc:108]: 53ec4cc

File: lib/seadsa/DsaCompleteCallGraph.cc [hunk 5]
```diff
@@ -387,8 +377,7 @@ bool CompleteCallGraphAnalysis::runOnModule(Module &M) {
 
       for (CallGraphNode *cgn : scc) {
         Function *fn = cgn->getFunction();
-        if (!fn || fn->isDeclaration() || fn->empty())
-          continue;
+        if (!fn || fn->isDeclaration() || fn->empty()) continue;
```
Comment (caballa) [inline lib/seadsa/DsaCompleteCallGraph.cc:380]: I think here we should use `m_specGraphInfo`

File: lib/seadsa/DsaCompleteCallGraph.cc [hunk 6]
```diff
@@ -413,8 +402,7 @@ bool CompleteCallGraphAnalysis::runOnModule(Module &M) {
       std::vector<CallGraphNode *> cgns = call_graph_utils::SortedCGNs(scc);
       for (CallGraphNode *cgn : cgns) {
         Function *fn = cgn->getFunction();
-        if (!fn || fn->isDeclaration() || fn->empty())
-          continue;
+        if (!fn || fn->isDeclaration() || fn->empty()) continue;
```
Comment (caballa) [inline lib/seadsa/DsaCompleteCallGraph.cc:405]: This is the other place where I think we should use `m_specGraphInfo`

**Comment by agurfinkel** [comment]
@AVassilev98 rework history. Use interactive git-rebase for that. 
Make sure that first line of each message fits within the length bound of github.
Squash and re-order commits that fix bugs in earlier commits.
If there is no reason for keeping history, squash everything into one commit.

File: include/seadsa/BottomUp.hh [hunk 7]
```diff
@@ -32,20 +33,24 @@ private:
   const llvm::DataLayout &m_dl;
   llvm::TargetLibraryInfoWrapperPass &m_tliWrapper;
   const AllocWrapInfo &m_allocInfo;
+  const SpecGraphInfo &m_specGraphInfo;
```
Comment (agurfinkel) [inline include/seadsa/BottomUp.hh:36]: Can we come up with a more descriptive name than `SpecGraphInfo`? 

I do not like the combination of three very generic terms: *Specification*, *Grap*, and *Information*

File: include/seadsa/CallSite.hh [hunk 8]
```diff
@@ -7,8 +7,8 @@
 #include "llvm/IR/Module.h"
 #include "llvm/Pass.h"
 
-// XXX: included for Cell. 
-#include "seadsa/Graph.hh" 
+// XXX: included for Cell.
```
Comment (agurfinkel) [inline include/seadsa/CallSite.hh:10]: there are a lot of formatting changes that are intermixed with code changes. 
They should be separated into separate commits. Maybe they are -- I am looking at this
through github interface that combines all the changes. Ideally formatting would be a separate PR to make reviewing easier.

File: include/seadsa/SpecGraphInfo.hh [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+class SpecGraphInfo : public llvm::ImmutablePass {
+
+protected:
+  mutable AllocWrapInfo *m_awi;
+  mutable llvm::TargetLibraryInfoWrapperPass *m_tliWrapper;
+  mutable std::unordered_map<std::string, llvm::Function *> m_funcs;
+
+  // m_module must be deallocated before m_ctx, keep in this order
+  mutable llvm::LLVMContext m_ctx;
+  mutable std::unique_ptr<llvm::Module> m_module;
```
Comment (agurfinkel) [inline include/seadsa/SpecGraphInfo.hh:40]: should not be hard to extend it to multiple modules. We might have a spec module per library rather than one module that links all specs together.

File: lib/seadsa/DsaCallSite.cc [hunk 10]
```diff
@@ -134,6 +114,20 @@ DsaCallSite::const_formal_iterator DsaCallSite::formal_end() const {
                                      getCallee()->arg_end());
 }
 
+DsaCallSite::const_formal_iterator
```
Comment (agurfinkel) [inline lib/seadsa/DsaCallSite.cc:117]: use `llvm::make_filter_range` instead of `boost::make_filter_iterator`

File: lib/seadsa/DsaCallSite.cc [hunk 11]
```diff
@@ -134,6 +114,20 @@ DsaCallSite::const_formal_iterator DsaCallSite::formal_end() const {
...
+DsaCallSite::formal_func_begin(const llvm::Function &F) const {
```
Comment (agurfinkel) [inline lib/seadsa/DsaCallSite.cc:118]: Comment why this is a method in `DsaCallSite`. It seems to work on a given function `F`, not necessarily either the caller nor the calee that the`DsaCallSite` represents

File: lib/seadsa/SpecGraphInfo.cc [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+void SpecGraphInfo::initialize() const {
+
+  using namespace llvm::sys::fs;
+  using namespace llvm;
+
+  m_awi = &getAnalysis<AllocWrapInfo>();
+  m_tliWrapper = &getAnalysis<TargetLibraryInfoWrapperPass>();
+
+  if (SpecFile.empty()) return;
```
Comment (agurfinkel) [inline lib/seadsa/SpecGraphInfo.cc:44]: should do this check before getting analysis since they might be computationally expensive.

File: lib/seadsa/SpecGraphInfo.cc [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+  auto &spec_funcs = m_module->getFunctionList();
+
+  for (llvm::Function &func : spec_funcs) {
+    if (func.isDeclaration() || func.empty()) continue;
+    m_funcs.insert({func.getName().str(), &func});
+  }
+} // namespace seadsa
+
+bool SpecGraphInfo::hasSpecFunc(const llvm::Function &F) const {
+  return m_funcs.count(F.getName().str());
```
Comment (agurfinkel) [inline lib/seadsa/SpecGraphInfo.cc:59]: I think you are using `std::string` here. Should avoid and use llvm specific strings, i.e., Twine or StringRef as appropriate

File: include/seadsa/CallSite.hh (hunk 8 above)
Comment (AVassilev98) [reply include/seadsa/CallSite.hh:10]: I have my ide set to run clang-format whenever I save, I'll make a separate commit from now on

Comment (agurfinkel) [reply include/seadsa/CallSite.hh:10]: you can fix this now by:
* checkout current `dev10`
* clang-format files you plan to change
* submit PR
* rebase your changes once the PR is accepted
* use meld as a mergetool to resolve any conflicts

Comment (agurfinkel) [reply include/seadsa/CallSite.hh:10]: This is what I would do, but we can also leave this as is since both @caballa and myself reviewed the code as it is already

File: lib/seadsa/DsaCallSite.cc (hunk 11 above)
Comment (AVassilev98) [reply lib/seadsa/DsaCallSite.cc:118]: I wasn't sure where else to put it. Would making a FunctionUtils file be more appropriate?

File: include/seadsa/SpecGraphInfo.hh (hunk 9 above)
Comment (AVassilev98) [reply include/seadsa/SpecGraphInfo.hh:40]: How would that manifest itself? would the user have to provide a directory where all specs are stored, a list of paths, something else?

File: lib/seadsa/SpecGraphInfo.cc (hunk 12 above)
Comment (AVassilev98) [reply lib/seadsa/SpecGraphInfo.cc:44]: Shall I also make a `isInitialized` flag to return if the SpecGraphInfo pass is already initialized?

File: include/seadsa/BottomUp.hh (hunk 7 above)
Comment (AVassilev98) [reply include/seadsa/BottomUp.hh:36]: I agree, but I struggled coming up with a name, since at its core the pass just grabs all the spec functions and puts them in a map. It's the other passes that actually have to do the work of using the spec information provided. Do you have any suggestions?

File: lib/seadsa/SpecGraphInfo.cc (hunk 12 above)
Comment (agurfinkel) [reply lib/seadsa/SpecGraphInfo.cc:44]: sounds good

File: lib/seadsa/DsaCallSite.cc (hunk 11 above)
Comment (agurfinkel) [reply lib/seadsa/DsaCallSite.cc:118]: if it is just a function and not a method, make it a function. There are lots of examples of functions that return ranges already. Don't make it member of `DsaCallSite`. Since the function is small, you can keep it in a header file.

File: include/seadsa/SpecGraphInfo.hh (hunk 9 above)
Comment (agurfinkel) [reply include/seadsa/SpecGraphInfo.hh:40]: I think the easiest is the option repeated many times as in `--dsa-spec=/ag/spec1.bc --dsa-spec=/ag/spec2.bc`.

For inspiration, think `-l` or `-L` flag for linking libraries in the compiler

File: include/seadsa/BottomUp.hh (hunk 7 above)
Comment (agurfinkel) [reply include/seadsa/BottomUp.hh:36]: `DsaLibFuncInfo` -- dsa-specific information for known library functions

**Comment by AVassilev98** [comment]
Separate commits just for ease of review. I'll squash everything down to a single commit once it's deemed good

**Comment by agurfinkel** [comment]
@AVassilev98 the CI build is failing. Looks like clang-format is not in the docker. Check, it is probably called `clang-format-10`

There should be some cmake magic to discover available version of `clang-format`

File: lib/seadsa/DsaGlobal.cc [hunk 14]
```diff
@@ -694,7 +694,8 @@ bool BottomUpGlobalAnalysis::runOnModule(Module &M) {
   LOG("dsa-global", errs() << "Started bottom-up global analysis ... \n");
 
   for (auto &F : M) {
-    if (F.isDeclaration() || F.empty()) continue;
+    if (F.empty() && !m_specGraphInfo.hasSpecFunc(F)) continue;
```
Comment (agurfinkel) [inline lib/seadsa/DsaGlobal.cc]: pull the common condition out

```
if (!m_specGraphInfo.hasSpecFunc(F)) {
     if (F.empty() || F.isDeclaration()) continue;
}
```

File: lib/seadsa/DsaGlobal.cc [hunk 15]
```diff
@@ -612,7 +612,8 @@ bool BottomUpTopDownGlobalAnalysis::runOnModule(Module &M) {
       errs() << "Started bottom-up + top-down global analysis ... \n");
 
   for (auto &F : M) {
-    if (F.isDeclaration() || F.empty()) continue;
+    if (F.empty() && !m_specGraphInfo.hasSpecFunc(F)) continue;
```
Comment (agurfinkel) [inline lib/seadsa/DsaGlobal.cc]: pull common condition out. See comment in another commit for the same looking code.

File: lib/seadsa/DsaLibFuncInfo.cc [hunk 16]
```diff
@@ -14,12 +14,23 @@
...
 #include "seadsa/support/Debug.h"
+#include <filesystem>
 #include <iostream>
 
 static llvm::cl::list<std::string>
     XSpecFiles("sea-dsa-specfile", llvm::cl::desc("<Input spec bitcode file>"),
                llvm::cl::ZeroOrMore, llvm::cl::value_desc("filename"));
 
+static llvm::cl::opt<bool>
+    XUse32BitSpec("sea-dsa-use-32bit-specs",
```
Comment (agurfinkel) [inline lib/seadsa/DsaLibFuncInfo.cc:24]: 32 vs 64 bit should be decided based on current `DataLayout`. You can grab a `DataLayout` from `Module` and check whether current pointer width is `32` or `64` bits

File: lib/seadsa/DsaLibFuncInfo.cc [hunk 17]
```diff
@@ -33,19 +44,43 @@ void DsaLibFuncInfo::initialize() const {
...
+
+  if (XUseClibSpec) {
+    if (XUse32BitSpec) {
+      const StringRef libcSpecRel = "../lib/libc-32.spec.bc";
+      llvm::SmallString<256> libcSpecAbs = libcSpecRel;
+      make_absolute(exeDir, libcSpecAbs);
+
+      m_modules.push_back(parseIRFile(libcSpecAbs.str(), err, m_ctx));
+    } else {
+      const StringRef libcSpecRel = "../lib/libc.spec.bc";
```
Comment (agurfinkel) [inline lib/seadsa/DsaLibFuncInfo.cc:65]: probably good idea to embed architecture in both spec files. We do sometimes run things in a 32 bit container, in which case, the architecture is assumed to be 32 bit by default.

File: lib/seadsa/DsaLibFuncInfo.cc [hunk 18]
```diff
@@ -33,19 +44,43 @@ void DsaLibFuncInfo::initialize() const {
...
+      llvm::SmallString<256> libcSpecAbs = libcSpecRel;
+      make_absolute(exeDir, libcSpecAbs);
+
+      m_modules.push_back(parseIRFile(libcSpecAbs.str(), err, m_ctx));
+    }
```
Comment (agurfinkel) [inline lib/seadsa/DsaLibFuncInfo.cc:70]: seems like this code does not expect any failure during parsing / reading

File: tools/CMakeLists.txt [hunk 19]
```diff
@@ -26,3 +33,35 @@ if (SEA_DSA_STATIC_EXE)
...
+
+add_custom_command(
+  OUTPUT ${SPEC_DIR}/libc.spec.c 
+  DEPENDS libcSpecGen 
+  COMMAND libcSpecGen
+  COMMAND clang-format -i ${SPEC_DIR}/libc.spec.c
+)
+
+add_library(LibcSpecs OBJECT ${SPEC_DIR}/libc.spec.c)
+set_target_properties(LibcSpecs PROPERTIES
```
Comment (agurfinkel) [inline tools/CMakeLists.txt:50]: you expect that everything is compiled with clang. This is not always the case. Perhaps a warning or at least a comment in the cmake file should be given in case someone compiles with gcc instead

File: tools/CMakeLists.txt [hunk 20]
```diff
@@ -26,3 +33,35 @@ if (SEA_DSA_STATIC_EXE)
...
+   DESTINATION lib
+   RENAME libc.spec.bc)
+
+option(SEAHORN_BUILD_32_BIT_SPEC "Build 32bit library specs" OFF)
+if(SEAHORN_BUILD_32_BIT_SPEC) 
+  add_library(LibcSpecs-32 OBJECT ${SPEC_DIR}/libc.spec.c)
+  set_target_properties(LibcSpecs-32 PROPERTIES
+    COMPILE_FLAGS "-emit-llvm -m32" LINK_FLAGS "-emit-llvm -m32")
+
+    install (FILES  ${CMAKE_CURRENT_BINARY_DIR}/CMakeFiles/LibcSpecs-32.dir/specs/libc.spec.c.o
```
Comment (agurfinkel) [inline tools/CMakeLists.txt:63]: file path seems wrong, here and in prior cases. I don't think we can make any assumption about the content of `CMAKE_CURRENT_BINARY_DIR`. CMake is free to rename `CMakeFiles` in the future. Usually `CMAKE_CURRENT_BINARY` dir is sufficient.

The main problem is that this code should be moved into CMakeLists file in `specs` directory

File: tools/libcSpecGen.cc [hunk 21]
```diff
@@ -0,0 +1,511 @@
+#include <fstream>
```
Comment (agurfinkel) [inline tools/libcSpecGen.cc:1]: add a comment with author, email, and what the file is expected to do.
This way we can contact you later if we run into problems ;)

File: tools/libcSpecGen.cc [hunk 22]
```diff
@@ -0,0 +1,511 @@
...
+#define NRET_NNNYARGS            \
+  {                              \
+    0, 0, 0, 0, 1, 1, 1, 1, 1, 1 \
+  }
+
+const struct
+{
+  const std::string name;
+  const libAction action;
+} recFuncs[] = {
```
Comment (agurfinkel) [inline tools/libcSpecGen.cc:122]: acknowledge where this table is from

File: lib/seadsa/DsaGlobal.cc (hunk 14 above)
Comment (AVassilev98) [reply lib/seadsa/DsaGlobal.cc]: I don't think this is the same logic. If the function is defined but and don't have a spec for it my code won't `continue` but yours will

File: tools/libcSpecGen.cc (hunk 22 above)
Comment (AVassilev98) [reply tools/libcSpecGen.cc:122]: would a link to the repo / file in repo suffice?

File: lib/seadsa/DsaLibFuncInfo.cc (hunk 18 above)
Comment (AVassilev98) [reply lib/seadsa/DsaLibFuncInfo.cc:70]: I'll add some error checking.

File: lib/seadsa/DsaCompleteCallGraph.cc (hunk 5 above)
Comment (caballa) [reply lib/seadsa/DsaCompleteCallGraph.cc:380]: I promised to do it after some cleanup/improvements I want to do.

File: lib/seadsa/DsaGlobal.cc [hunk 23]
```diff
@@ -193,17 +191,18 @@ bool ContextInsensitiveGlobalPass::runOnModule(Module &M) {
   auto &dl = M.getDataLayout();
   auto &tli = getAnalysis<TargetLibraryInfoWrapperPass>();
   auto &allocInfo = getAnalysis<AllocWrapInfo>();
+  auto &dsaLibFuncInfo = getAnalysis<DsaLibFuncInfo>();
```
Comment (caballa) [inline lib/seadsa/DsaGlobal.cc:194]: Don't we need to initialize here `dsaLibFuncInfo` ?

File: lib/seadsa/DsaGlobal.cc [hunk 24]
```diff
@@ -224,11 +224,12 @@ bool FlatMemoryGlobalPass::runOnModule(Module &M) {
   auto &tli = getAnalysis<TargetLibraryInfoWrapperPass>();
   auto &cg = getAnalysis<CallGraphWrapperPass>().getCallGraph();
   auto &allocInfo = getAnalysis<AllocWrapInfo>();
-  allocInfo.initialize(M, this);  
+  auto &dsaLibFuncInfo = getAnalysis<DsaLibFuncInfo>();
```
Comment (caballa) [inline lib/seadsa/DsaGlobal.cc:227]: again shouldn't we call `initialize` ?

File: lib/seadsa/DsaLibFuncInfo.cc (hunk 18 above)
Comment (caballa) [reply lib/seadsa/DsaLibFuncInfo.cc:76]: The code for 32 and 64 is exactly the same except lines 55 and 66. We can factorize code a bit

File: include/seadsa/DsaLibFuncInfo.hh [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+class AllocWrapInfo;
+
+class DsaLibFuncInfo : public llvm::ImmutablePass {
+  using ModuleRef = std::unique_ptr<llvm::Module>;
+
+protected:
+  mutable bool isInitialized = false;
+  mutable AllocWrapInfo *m_awi;
+  mutable llvm::TargetLibraryInfoWrapperPass *m_tliWrapper;
+  mutable std::unordered_map<std::string, llvm::Function *> m_funcs;
```
Comment (caballa) [inline include/seadsa/DsaLibFuncInfo.hh:38]: Maybe better a `StringMap` ?

Comment (AVassilev98) [reply include/seadsa/DsaLibFuncInfo.hh]: 160ef53

File: lib/seadsa/DsaGlobal.cc (hunk 24 above)
Comment (AVassilev98) [reply lib/seadsa/DsaGlobal.cc:247]: 3458679
